### PR TITLE
add q as built-in

### LINKED ISSUE #412: [Bug]: q is not a built-in function
### What version of Datahike are you using?

0.3.7-SNAPSHOT

### What version of Java are you using?

openjdk version "11.0.11" 2021-04-20

### What operating system are you using?

Ubuntu 20.04.3 LTS

### What database EDN configuration are you using?

``` edn
{:store              {:backend :mem
                      :id      "sandbox" }
 :name               "sandbox"
 :keep-history?      true
 :schema-flexibility :write
 :attribute-refs?    true}
```

### Describe the bug

Datahike does not support the build-in function `q` inside q while in Datomic it is [supported](https://docs.datomic.com/cloud/query/query-data-reference.html#q).

### What is the expected behaviour?

It should be possible to use `q` as a built-in function.

### How can the behaviour be reproduced?

```clojure
(def schema [{:db/ident       :name
              :db/cardinality :db.cardinality/one
              :db/index       true
              :db/unique      :db.unique/identity
              :db/valueType   :db.type/string}
             {:db/ident       :parents
              :db/cardinality :db.cardinality/many
              :db/valueType   :db.type/ref}
             {:db/ident       :age
              :db/cardinality :db.cardinality/one
              :db/valueType   :db.type/long}])

(def cfg {:store              {:backend :mem
                               :id      "sandbox" }
          :name               "sandbox"
          :keep-history?      true
          :schema-flexibility :write
          :attribute-refs?    true})

(d/delete-database cfg)

(d/create-database cfg)

(def conn (d/connect cfg))

(d/transact conn {:tx-data schema})

(d/transact conn {:tx-data [{:name "Alice"
                             :age  35}
                            {:name "Bob"
                             :age  37}]})

(d/q '[:find ?n ?a ?t
       :where
       [(q '[:find (min ?a) :where [_ :age ?a]] $) [[?a]]]
       [?e :age ?a]
       [?e :name ?n ?t]]
     @conn)
;; => Execution error (ExceptionInfo) at datahike.query/bind-by-fn (query.cljc:694).
;; Unknown function 'q in [(q (quote [:find (min ?a) :where [_ :age ?a]]) $) [[?a]]]

```

### PR DESCRIPTION
**Pull request by kordano**
#### SUMMARY
Fixes #412 

##### Bugfix
- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [x] Formatting checked

### CODE REVIEW

File: src/datahike/query.cljc [hunk 1]
```diff
@@ -1097,25 +1102,27 @@
...
+        find          (:qfind parsed-q)
         find-elements (dpip/find-elements find)
-        find-vars (dpi/find-vars find)
-        result-arity (count find-elements)
-        with (:qwith parsed-q)
-        returnmaps (:qreturnmaps parsed-q)
+        find-vars     (dpi/find-vars find)
+        result-arity  (count find-elements)
+        with          (:qwith parsed-q)
+        returnmaps    (:qreturnmaps parsed-q)
```
Comment (jsmassa) [inline src/datahike/query.cljc:1125]: As far as I know, we agreed to avoid unnecessary whitespace changes like these to keep the PRs clear and to the point.  But since this is such a small PR I'd say it's ok here.

**Review by jsmassa — APPROVED** [review]
Nice :)

File: src/datahike/query.cljc (hunk 1 above)
Comment (kordano) [reply src/datahike/query.cljc:1125]: Ok, we should anyway talk about formatting and style guide. I'll create a discussion for that.
